Stats Traffic Charts

ravishanker opened the pull request:
https://github.com/wordpress-mobile/jetpack-issue-repo/issues/14#issue-2147025725

Adds bar graphs for Week, Month, and Year granularity

| Day | Week | Month | Year |
|--------|--------|--------|--------|
| ![Screenshot_20240220_145219](https://github.com/wordpress-mobile/WordPress-Android/assets/990349/ee5079e8-9768-4782-8568-c8dca40b564f) | ![Screenshot_20240223_163324](https://github.com/wordpress-mobile/WordPress-Android/assets/990349/c2414551-d50a-47b8-ae3a-5f787b839f14) |![Screenshot_20240223_163355](https://github.com/wordpress-mobile/WordPress-Android/assets/990349/f7b93015-d209-46b4-b7cf-7f10e527c148) | ![Screenshot_20240223_163417](https://github.com/wordpress-mobile/WordPress-Android/assets/990349/59af2554-931d-4cb7-9ee6-8ed9eb0d32cb) | 

-----

## To Test:

Test 1:
- Enable `stats_traffic_tab` feature flag (Me -> Debug settings - Remote Features)
- Go to Stats
- `Verify` only Traffic tab, and Insights tabs are shown as in after column 
- `Switch` to Traffic tab if not on it already
- `Select` By day from the drop-down
- `Verify` that Today card is shown in the image above
- `Select` By week from the drop-down
- `Verify` a Bar chart is shown with seven bars for 7 days in a week

> [!NOTE] 
> This is still work in progress.  Date selector, and graphs need to be in sync
> You may also check `By month`, and `By year` options from the drop-down

Test 2:

- Disable feature flag
- `Verify` Stats tabs all show up, and work as before



-----

## Regression Notes

1. Potential unintended areas of impact

    - Behind feature flag

2. What I did to test those areas of impact (or what existing automated tests I relied on)

    - Tested 

3. What automated tests I added (or what prevented me from doing so)

    - Updated existing unit tests

-----

## PR Submission Checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

-----

## Testing Checklist:

- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

ravishanker committed:
Create TrafficOverviewMapper.kt

Add Mapper for Traffic Charts

ravishanker committed:
Update strings.xml

ravishanker committed:
Create stats_block_traffic_value_item.xml

ravishanker committed:
Create TrafficValueViewHolder.kt

ravishanker committed:
Create stats_block_traffic_selectable_column.xml

ravishanker committed:
Create stats_block_traffic_four_columns_item.xml

ravishanker committed:
Create TrafficFourColumnsViewHolder.kt

ravishanker committed:
Update StatsViewAllFragment.kt

ravishanker committed:
Update LoadingViewHolder.kt

ravishanker committed:
Update BlockListViewHolder.kt

ravishanker committed:
Update StatsListFragment.kt

ravishanker committed:
Update StatsBlockAdapter.kt

ravishanker committed:
Update BlockListItem.kt

ravishanker committed:
Create TrafficBarChartViewHolder.kt

ravishanker committed:
Update BlockListAdapter.kt

ravishanker committed:
Update OverviewUseCase.kt

ravishanker committed:
Update OverviewUseCaseTest.kt

ravishanker committed:
Update ViewsWidgetListViewModelTest.kt

dangermattic commented:
<!--
  0 Errors
  7 Warnings: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          7 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>TrafficOverviewMapper</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>SelectedType</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>TrafficOverviewUseCase</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>UiState</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>TrafficOverviewUseCaseFactory</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>BarChartLabelFormatter</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

ravishanker committed:
Create stats_block_traffic_bar_chart_item.xml

ravishanker committed:
Update TrafficBarChartViewHolder.kt

ravishanker committed:
Merge branch 'trunk' into Stats-Traffic-By-Week-Chart

ravishanker committed:
Create BarChartLabelFormatter.kt

ravishanker committed:
Update TrafficBarChartViewHolder.kt

ravishanker committed:
Update TrafficOverviewMapper.kt

ravishanker committed:
Update StatsDateFormatter.kt

irfano committed:
Merge branch 'trunk' into Stats-Traffic-By-Week-Chart

irfano committed:
Update granularity usage in OverviewUseCase

ravishanker committed:
Create TrafficOverviewUseCase.kt

Separate TrafficUseCase from previous OverviewUseCase

ravishanker committed:
Update OverviewUseCase.kt

Remove traffic tab related conditions as a new traffic overview use case is created

ravishanker committed:
Update StatsModule.kt

Add TrafficOverviewUseCase

ravishanker committed:
Update StatsListViewModel.kt

ravishanker committed:
Remove title on traffic bar chart

ravishanker committed:
Update BlockListAdapter.kt

remove title, and update to four columns with values

ravishanker committed:
Delete unused

Delete unused ViewHolders, and layout files meant for title, and columns on traffic bar chart

ravishanker committed:
Update TrafficBarChartViewHolder.kt

adjust offset above columns

ravishanker committed:
Update TrafficBarChartViewHolder.kt

fix labels on Year bar chart to appear on every bar, not alternate bars

irfano committed:
Fix a crash in `LineChartLabelFormatter`

irfano commented:
`LineChartLabelFormatter` was crashing on my local. The reason might be the time zone difference between my home and the site. I fixed it: https://github.com/wordpress-mobile/WordPress-Android/pull/20188/commits/82c142dd8c3b08d39300e4c8040be72b8aa9a2de

notandyvee commented on WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/traffic/TrafficBarChartViewHolder.kt:
```diff
@@ -0,0 +1,251 @@
+
+    private fun configureYAxis(item: BlockListItem.BarChartItem) {
+        val minYValue = 4f
+        val maxYValue = item.entries.maxByOrNull { it.value }!!.value
```
I'd avoid the `!!`.

irfano replied:
Good catch! I got a crash caused by this. Fixed: b158c4f137cf91bf3878921f97f141cd875137c7

notandyvee commented on WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/traffic/TrafficBarChartViewHolder.kt:
```diff
@@ -0,0 +1,251 @@
+        count: Int,
+        entries: List<T>
+    ): List<T> {
+        return if (count < entries.size) entries.subList(
```
You should be able to get rid of the function block parens and just do

> ): List<T> = if (count < entries.size) entries.subList(...

notandyvee commented on WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/traffic/TrafficBarChartViewHolder.kt:
```diff
@@ -0,0 +1,251 @@
+    }
+
+    private fun toBarEntry(bar: Bar, index: Int): BarEntry {
+        return BarEntry(
```
Same here

notandyvee commented on WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/traffic/TrafficBarChartViewHolder.kt:
```diff
@@ -0,0 +1,251 @@
+
+    @Suppress("ReturnCount")
+    private fun roundUp(input: Float): Float {
+        return if (input > 100) {
```
Same. I'll stop pointing it out as there are other places too. Just makes things look cleaner.

notandyvee commented on WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/traffic/TrafficOverviewUseCase.kt:
```diff
@@ -0,0 +1,314 @@
+            val lastDayDate = statsDateFormatter.parseStatsDate(granularity, lastDayData.period)
+            if (lastDayDate.before(yesterday.time)) {
+                val currentCalendar = localeManagerWrapper.getCurrentCalendar()
+                val lastItemAge = ceil((currentCalendar.timeInMillis - lastDayDate.time) / 86400000.0)
```
`86400000.0` definitely needs some explanation. A constant could help explain why.

irfano replied:
This `logIfIncorrectData` fun has been copied from other places so it's not new. I'm unsure if this issue is still valid and we should continue tracking it.

notandyvee commented on WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/utils/LineChartLabelFormatter.kt:
```diff
@@ -10,7 +10,7 @@ class LineChartLabelFormatter @Inject constructor(
     override fun getAxisLabel(value: Float, axis: AxisBase?): String {
         val index = value.toInt()
-        return if (index < entries.size) {
+        return if (entries.isNotEmpty() && index < entries.size) {
```
What if `val index = value.toInt()` is -1? A range could fill this well I think.

`if (index in 1..entries.size) {`

This works because the [rangeTo](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.ranges/range-to.html) docs state:

> This value needs to be smaller than or equal to [that](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.ranges/range-to.html#kotlin.ranges$rangeTo(kotlin.ranges.rangeTo.T,%20kotlin.ranges.rangeTo.T)/that) value, otherwise the returned range will be empty.

Also, you add the not empty check here, but not in `BarChartLabelFormatter`.

notandyvee commented on WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/traffic/TrafficBarChartViewHolder.kt:
```diff
@@ -0,0 +1,251 @@
+import org.wordpress.android.ui.stats.refresh.utils.LargeValueFormatter
+import org.wordpress.android.util.DisplayUtils
+
+@Suppress("MagicNumber")
```
Some of the numbers below could use a constant name to help identify and explain what they do. Might be worth removing this and fixing for clarity.

ravishanker replied:
Normally yes, I would've said the same.  However, in this case it is a graph, they are just property settings.

notandyvee reviewed (commented):
Used the review as opportunity to learn and understand the code. Some small comments mostly style. Looking good.

irfano committed:
Merge branch 'trunk' into Stats-Traffic-By-Week-Chart

irfano committed:
Handle nullability in TrafficBarChartViewHolder

irfano committed:
Allow TrafficOverviewUseCase to fetch when the date is changed

irfano committed:
Update endpoint usage within `TrafficOverviewUseCase` to fill the view

irfano commented:
The date selector, overview card, and chart are working as expected after my latest commits. Now, we can see and test all chart types in this PR.
- I think we can address Andy's comments, and
- update the UI of the card with the latest design (p1708598484003149/1708374772.289219-slack-C06BR07TJHK),

then close this PR. There is an issue with loading the previous dates; the cards display wrong values for a very short time before showing the correct value. I'll look at it in a separate PR.

ravishanker committed:
Update LineChartLabelFormatter.kt

refactor to use range

ravishanker committed:
Update BarChartLabelFormatter.kt

refactor to use range

ravishanker committed:
Update TrafficBarChartViewHolder.kt

refactor function block to expression body

ravishanker committed:
Update BarChartLabelFormatter.kt

update range to start at 0

ravishanker committed:
Update LineChartLabelFormatter.kt

update range to start at 0

ravishanker committed:
Update TrafficOverviewUseCase.kt

remove redundant logging

ravishanker committed:
Revert "Delete unused"

This reverts commit f0b05c29d75844df8b02e887a304ccdce9583864.

ravishanker committed:
Update stats_styles.xml

ravishanker committed:
Update stats_block_traffic_selectable_column.xml

ravishanker committed:
Update TrafficFourColumnsViewHolder.kt

ravishanker committed:
Update BlockListAdapter.kt

ravishanker committed:
Update TrafficBarChartViewHolder.kt

ravishanker committed:
Delete TrafficValueViewHolder.kt

ravishanker committed:
Delete stats_block_traffic_value_item.xml

ravishanker committed:
Update OverviewUseCaseTest.kt

ravishanker commented:
> then close this PR. There is an issue with loading the previous dates; the cards display wrong values for a very short time before showing the correct value. I'll look at it in a separate PR.

Nice work 🥇   Sounds good 👍.  There could also be some stylistic issues that can be fixed in separate PR after Chris review.

irfano committed:
Remove unused resources and code for Overview card title

irfano reviewed (approved):
The change LGTM! I'm approving the PR, but I removed `Fixes #19951` from the description since the issue needs 2 more tasks:

- [ ] Fix the crash occurs when the Talkback is enabled. I have already started this task. I will fix the crash and test accessibility.
- [ ] Design updates from the notes I explained in the screenshot.

![emulator-vs-figma](https://github.com/wordpress-mobile/WordPress-Android/assets/2471769/f3c0b2bb-4720-4352-92c2-1567650df4ee)

ravishanker committed:
Update TrafficBarChartViewHolder.kt

Update color to match designs

ravishanker committed:
Adjust padding on bar graph

ravishanker commented:
> The change LGTM! I'm approving the PR, but I removed `Fixes #19951` from the description since the issue needs 2 more tasks:

> * [x]  Design updates from the notes I explained in the screenshot.
> 
Updated the color, the margin is required, otherwise the Y axis values are being cut of higher value range, probably it's looking better without being edge-to-edge.
